src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java

```diff
@@ -33,46 +63,114 @@
+        if (intervals.contains(null)) {
+            throw new IllegalArgumentException("the input cannot contain null");
+        }
+
+        intervalOverlapComparator = buildIntervalOverlapComparator(contigComparator);
+        intervalStandardComparator = buildIntervalStandardComparator(contigComparator);
+        sortedIntervals = intervals.stream().sorted(intervalStandardComparator).collect(Collectors.toList());
+        checkForOverlaps(sortedIntervals,intervalOverlapComparator);
+        this.intervalsByName = composeIntervalsByName(sortedIntervals);
     }
 
     /**
-     * Throws a exception if there some interval location incoherence.
+     * Fails with an exception if the intervals collection has overlapping intervals.
+     * @param sortedIntervals
+     * @param overlapComparator
+     */
+    private static void checkForOverlaps(final List<SimpleInterval> sortedIntervals,
+                                  final Comparator<SimpleInterval> overlapComparator) {
+        final OptionalInt failureIndex = IntStream.range(1, sortedIntervals.size())
+                .filter(i -> overlapComparator.compare(sortedIntervals.get(i-1),sortedIntervals.get(i)) == 0)
+                .findFirst();
+
+        if (failureIndex.isPresent()) {
+            final int index = failureIndex.getAsInt();
+            throw new IllegalArgumentException(
+                    String.format("input intervals contain at least two overlapping intervals: %s and %s",
+                            sortedIntervals.get(index-1),sortedIntervals.get(index)));
+        }
+    }
+
+    /**
+     * Given a contig ({@link String}) comparator it will compose interval comparator that will
+     * return 0 when both intervals are overlapping.
      *
      * <p>
-     *     These include:
-     *     <ul>
-     *         <li>{@code} null intervals,</li>
-     *         <li>intervals are out of order,</li>
-     *         <li>intervals don't overlap each other,</li>
-     *         <li>some interval is unmapped or</li>
-     *         <li>is the special value {@link GenomeLoc#WHOLE_GENOME}</li>
-     *     </ul>
+     *     See {@link #intervalOverlapComparator} for more details.
      * </p>
+     * @param contigComparator the contig comparator.
+     * @return never {@code null}.
+     */
+    private Comparator<SimpleInterval> buildIntervalOverlapComparator(final Comparator<String> contigComparator) {
+
+        // It turns out that figuring out if there is an overlap is simply whether
+        // the cross-comparisons of start and end positions of both intervals is contradictory:
+        return Comparator.comparing(SimpleInterval::getContig, contigComparator)
+                .thenComparing((a, b) -> {
+                    final int startEndCompare = Integer.compare(a.getStart(),b.getEnd());
+                    final int endStartCompare = Integer.compare(a.getEnd(),b.getStart());
+                    return startEndCompare == endStartCompare ? startEndCompare : 0;
```

Comment:
Would this be clearer?  
`` a.overlaps(b) : 0 ? Integer.compare(a.getStart(), b.getStart()``